### PR TITLE
[Model][LoRA]LoRA support added for MiniCPMV2.6

### PR DESCRIPTION
**Pull request by jeejeelee**
ping @DarkLight1337 
cc @HwwwwwwwH   I replaced 2.6's vision encoder, could you  please help check if there are any issues? Although I've already tested it locally

### CODE REVIEW

File: vllm/model_executor/models/minicpmv.py [hunk 1]
```diff
@@ -960,12 +981,12 @@ def get_vision_hidden_states(
...
             patch_attention_mask=patch_attn_mask,
             tgt_sizes=tgt_sizes,
-        ).last_hidden_state
+        )
 
         return self.resampler(vision_embedding, tgt_sizes)
 
     def is_default_weight_loading(self, name: str) -> bool:
-        return "resampler" in name or "vpm" in name
+        return "resampler" in name  #or "vpm" in name
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:989]: Is this intended?

Comment (jeejeelee) [reply vllm/model_executor/models/minicpmv.py:989]: Yep, this commented code should be deleted

Comment (DarkLight1337) [reply vllm/model_executor/models/minicpmv.py:989]: So the original code is wrong?

Comment (jeejeelee) [reply vllm/model_executor/models/minicpmv.py:989]: similar to :https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/minicpmv.py#L868

**Comment by DarkLight1337** [comment]
I tested the output on the example script and found slightly different results. But that may be because of the TP'ed attention module.

File: vllm/model_executor/models/minicpmv.py (hunk 1 above)
Comment (jeejeelee) [reply vllm/model_executor/models/minicpmv.py:989]: We tested this modified minicpmv26 locally using script  similar to [test_minicpmv](https://github.com/vllm-project/vllm/blob/main/tests/models/decoder_only/vision_language/test_minicpmv.py), and the test passed correctly. For the LoRA test, we trained a LoRA using our internal data, and the inference worked without issues.

Comment (DarkLight1337) [reply vllm/model_executor/models/minicpmv.py:936]: Actually let me remove the comment by myself

**Review by DarkLight1337 — APPROVED** [review]
LGTM after testing, thanks again!

**Comment by jeejeelee** [comment]
@DarkLight1337  You're very kind, and I truly appreciate your hard work. Thank you so much.

**Comment by whyiug** [comment]
After this submission, I observed some differences. On my finetuned model(`modules_to_save = ['embed_tokens','resampler']`), there are differences in the output vision embedding.

This is my print debugging code.
```python
            vision_hidden_states = self.get_vision_hidden_states(image_inputs)
            print(f"shape of vision_hidden_states: {vision_hidden_states.shape}")
            print(f"preview of vision_hidden_states: {vision_hidden_states[0, 0, :5]}")
```

output before this PR:
```
shape of vision_hidden_states: torch.Size([9, 64, 3584])
preview of vision_hidden_states: tensor([-2.3418,  2.3047, -3.0117,  0.0443,  1.4688], device='cuda:0',
       dtype=torch.float16)
```
after:
```
shape of vision_hidden_states: torch.Size([9, 64, 3584])
preview of vision_hidden_states: tensor([-2.2266,  2.2031, -3.0723,  0.1958,  1.4131], device='cuda:0',
       dtype=torch.float16)
```

Also observed that the inference performance was relatively improved after this commit. Can you help me to check those issues? @jeejeelee

**Comment by jeejeelee** [comment]
This PR modified the implementation of the visual encoder, which could be the reason. Can you provide the code to reproduce this issue?

**Comment by whyiug** [comment]
> This PR modified the implementation of the visual encoder, which could be the reason. Can you provide the code to reproduce this issue?
@jeejeelee 
Just use the officially provided code and model (through finetuned for more differences)

env:
```
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: version 3.30.1
Libc version: glibc-2.17

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.105.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 545.23.08
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.1
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                12
On-line CPU(s) list:   0-11
Thread(s) per core:    2
Core(s) per socket:    6
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Platinum 8269CY CPU @ 2.50GHz
Stepping:              7
CPU MHz:               2499.976
BogoMIPS:              4999.95
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              36608K
NUMA node0 CPU(s):     0-11
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc eagerfpu pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single rsb_ctxsw fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat avx512_vnni

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-ml-py==12.555.43
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] nvidia-nvjitlink-cu12==12.5.82
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] optree==0.12.1
[pip3] pyzmq==26.0.3
[pip3] torch==2.4.0
[pip3] torchvision==0.19.0
[pip3] transformers==4.45.2
[pip3] triton==3.0.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-ml-py              12.555.43                pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.5.82                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] optree                    0.12.1                   pypi_0    pypi
[conda] pyzmq                     26.0.3                   pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torchvision               0.19.0                   pypi_0    pypi
[conda] transformers              4.45.2                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.6.4.dev19+gcf1d62a6.d20241016
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      0-11    0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```
```python
from transformers import AutoTokenizer

from vllm import LLM, SamplingParams
from PIL import Image
import requests


model_id = "openbmb/MiniCPM-V-2_6"
image_url = "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg"
image = Image.open(requests.get(image_url, stream=True).raw).convert("RGB")

question = "描述这张图片。"
tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True)
llm = LLM(
    model=model_id,
    trust_remote_code=True,
    max_model_len=2048,
    gpu_memory_utilization=0.9,
    max_num_seqs=5,
    dtype="auto",
)

messages = [{"role": "user", "content": f"(<image>./</image>)\n{question}"}]
prompt = tokenizer.apply_chat_template(
    messages, tokenize=False, add_generation_prompt=True
)
stop_tokens = ["<|im_end|>", "<|endoftext|>"]
stop_token_ids = [tokenizer.convert_tokens_to_ids(i) for i in stop_tokens]
sampling_params = SamplingParams(
    stop_token_ids=stop_token_ids,
    max_tokens=100,
    temperature=0,
    best_of=1,
)


llm_inputs = {"prompt": prompt, "multi_modal_data": {"image": image}}

outputs = llm.generate(
    [llm_inputs],
    sampling_params=sampling_params,
)
print(outputs[0].outputs[0].text)

```

output before this PR:
```
shape of vision_hidden_states: torch.Size([9, 64, 3584])
preview of vision_hidden_states: tensor([-1.7344,  0.7695, -1.9062, -0.5352,  0.5664], device='cuda:0',
       dtype=torch.bfloat16)
```
after:

```
shape of vision_hidden_states: torch.Size([9, 64, 3584])
preview of vision_hidden_states: tensor([-1.7031,  0.7578, -1.9453, -0.5117,  0.5469], device='cuda:0',
       dtype=torch.bfloat16)
```
That's still quite a difference, right?

**Comment by jeejeelee** [comment]
@whyiug   As for this output of yours, I think this is an acceptable difference. 

BTW, We tested this modified minicpmv26 locally using script similar to [test_minicpmv](https://github.com/vllm-project/vllm/blob/main/tests/models/decoder_only/vision_language/test_minicpmv.py), and the test passed correctly.

**Comment by whyiug** [comment]
I used the minicpmv model for the classification task and this difference between  `[-1.7344,  0.7695, -1.9062, -0.5352,  0.5664]` and `[-1.7031,  0.7578, -1.9453, -0.5117,  0.5469]` just led to exactly the opposite result:(

Thanks for the reply, I'll check the test script later.

**Comment by whyiug** [comment]
@jeejeelee BTW, can you explain in a little more detail why this PR( i mean the implementation of the visual encoder) get a big performance increase. Thanks.

**Comment by jeejeelee** [comment]
@whyiug  I haven't specifically tested the performance improvement of the visual encoder, but I think  the `QKVParallelLinear` (fused kernel )and the attention using xformer (optimized attention implementation)are the reasons.

**Pull request merged**